### PR TITLE
[11.0] product_internal_reference_mandatory

### PR DESCRIPTION
**Pull request by agyamuta**
Correspondence in reference to issue https://github.com/OCA/product-attribute/issues/331

### CODE REVIEW

**Comment by max3903** [comment]
@agyamuta Same comment as in https://github.com/OCA/product-attribute/pull/333#issuecomment-369245794

**Comment by max3903** [comment]
@agyamuta You need to define a `pre_init_hook` to load the data file before installation. See at the bottom of this page:
https://www.odoo.com/documentation/11.0/reference/module.html

**Comment by agyamuta** [comment]
Thanks for the comment @max3903. I have the following issues in using pre_init_hook.
a. With the current setup, module is first attempted to be installed prior to load the demo data. Installation failed  due to the existing data thats violates Null data for internal reference.
b. In the pre_init_hook, how to identify that the DB instance is configured to load Demo Data.
image

**Comment by max3903** [comment]
@agyamuta You can check in the module table on one of the dependency. There is a boolean field `demo`.

**Review by bodedra — APPROVED** [review]
LGTM. Thanks.

**Review by max3903 — APPROVED** [review]

File: product_code_mandatory/__init__.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+from . import models
+
+
+def pre_init_product_code(cr):
+    cr.execute("""UPDATE product_template
+        SET default_code = 'DEFAULT' || nextval('ir_default_id_seq')
+        WHERE default_code is NULL
+        OR LENGTH(default_code) = 0""")
+    cr.execute("""UPDATE product_product
```
Comment (feketemihai) [inline product_code_mandatory/__init__.py:12]: I would update the product.product with a reference from template, let's say the product template has DEFAULT100, and on variants associated you get DEFAULT100-1, DEFAULT100-2, when you search by template code on products, to have the variantss from it...

File: product_code_mandatory/README.rst [hunk 2]
```diff
@@ -0,0 +1,62 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+
+======================================================
+Set as Mandatory field for  Product Internal Reference
```
Comment (feketemihai) [inline product_code_mandatory/README.rst]: Use module name here..

File: product_code_mandatory/README.rst [hunk 3]
```diff
@@ -0,0 +1,62 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
```
Comment (feketemihai) [inline product_code_mandatory/README.rst]: Update with png image

File: product_code_mandatory/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# Copyright (C) 2018 - TODAY, Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Product Internal Reference as Mandatory",
```
Comment (feketemihai) [inline product_code_mandatory/__manifest__.py]: Use Required instead on Mandatory

File: product_code_mandatory/models/product.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Copyright (C) 2018 - TODAY, Open Source Integrators License AGPL-3.0
+# or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ProductProduct(models.Model):
```
Comment (feketemihai) [inline product_code_mandatory/models/product.py:7]: by default you create a product template, and you can add variants, the pre_init_hook is adding a default code to templates first, so here i think you should use the same, the product.template model.

**Review by feketemihai — CHANGES_REQUESTED** [review]

File: product_code_mandatory/__init__.py (hunk 1 above)
Comment (agyamuta) [reply product_code_mandatory/__init__.py:12]: @feketemihai thanks for the comment. 

Let me quote Max' comment:
We don't need to touch product_template as the default_code of a product template with many variants will be empty:

https://github.com/odoo/odoo/blob/11.0/addons/product/models/product_template.py#L256-L266

With Odoo's default behavior when a product.template is created it creates a product.product record in which the product.product.default_code is initialized with a unique value. This value then is pushed back to product.template.default_code.

Comment (max3903) [reply product_code_mandatory/__init__.py:12]: @agyamuta Be careful my comment was referring to the case of uniqueness. In this case, the field cannot be empty.

File: product_code_mandatory/models/product.py (hunk 5 above)
Comment (agyamuta) [reply product_code_mandatory/models/product.py:7]: The pre_init_hook process in this module is to make sure that all existing records(product_product table) would have an initial value and not empty or blank. This is only done once upon installation of the module.

**Comment by max3903** [comment]
@feketemihai Can you update your review?

File: product_code_mandatory/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    "category": "Product",
+    "website": "http://www.opensourceintegrators.com",
+    "depends": ["product"],
+    "data": [
+        "views/product_view.xml",
+        "views/product_code_seq.xml",
+    ],
+    'demo': [
+    ],
+    "pre_init_hook": 'pre_init_product_code',
```
Comment (MeetKD) [inline product_code_mandatory/__manifest__.py]: Can you please remove empty key ?

**Pull request merged**